### PR TITLE
coinbase transactions

### PR DESCRIPTION
**Pull request by harsh-ps-2003**
- [x] issue #173 
- [x] follows contribution [guide](https://github.com/keep-starknet-strange/shinigami/blob/main/CONTRIBUTING.md)
- [x] code change includes tests

Adds support for Coinbase Transactions (Basic and BIP34) with tests.
Also added optimised shift left and right operations as utility methods!

### CODE REVIEW

**Comment by harsh-ps-2003** [comment]
@b-j-roberts can you take a look?

File: src/test_coinbase.cairo [hunk 1]
```diff
@@ -0,0 +1,98 @@
+use shinigami::transaction::TransactionTrait;
+use shinigami::transaction::Transaction;
+use shinigami::transaction::TransactionOutput;
+use shinigami::transaction::TransactionInput;
```
Comment (b-j-roberts) [inline src/test_coinbase.cairo:4]: Can you combine these into something like `use shinigami::transaction::{TransactionTrait, Transaction, ...}`?

File: src/test_coinbase.cairo
Comment (b-j-roberts) [inline src/test_coinbase.cairo:1]: Could you move this file into `src/tests/`?

**Review by b-j-roberts — CHANGES_REQUESTED** [review]
Nice work so far!

I was thinking we could create a validate type function that takes a transaction argument and verifies it is a valid coinbase transaction.

I think the coinbase logic from this file should be a good guide on that : https://github.com/btcsuite/btcd/blob/2b53ed198955ac40fe5b3ce4a854e9f5bfa68258/blockchain/validate.go#L28

Lmk if you are able to do this. Thanks

File: src/transaction.cairo [hunk 2]
```diff
@@ -135,6 +202,31 @@ pub impl TransactionImpl of TransactionTrait {
...
+
+    fn calculate_block_subsidy(block_height: u32) -> i64 {
+        let halvings = block_height / 210000;
+        if halvings >= 64 {
+            return 0;
+        }
+        utils::shr::<i64, u32>(50 * 100000000, halvings)
+    }
+
+    fn validate_coinbase(tx: Transaction) -> bool {
```
Comment (b-j-roberts) [inline src/transaction.cairo:214]: Can you make this function return a `Result<(),felt252>`, and add some errors to `src/errors.cairo` for invalid coinbase?

**Review by b-j-roberts — CHANGES_REQUESTED** [review]
Would be nice to have a set of tests that create and validate various coinbase transactions from mainnet blocks.
Some that would be interesting to add would be the block 0 coinbase transaction, block 1 coinbase transaction, block 227,836 coinbase transaction ( 1st bip34 coinbase ), 481,824 coinbase transaction ( last before segwit ).

**Comment by harsh-ps-2003** [comment]
@b-j-roberts in case of testing using coinbase transactions from main net, I have only taken [one latest output instead of many](https://btcscan.org/block/000000000000000000cbeff0b533f8e1189cf09dfbebf57a8ebe349362811b80).

**Comment by harsh-ps-2003** [comment]
@b-j-roberts in case of testing using coinbase transactions from main net, I have only taken [one latest output instead of many](https://btcscan.org/block/000000000000000000cbeff0b533f8e1189cf09dfbebf57a8ebe349362811b80).

**Review by b-j-roberts — APPROVED** [review]
Thank you!
I added some extra features on top, like raw transaction decoding, to improve the tests for Coinbase a bit.

**Pull request merged**